frontend/catalyst/python_interface/transforms/quantum/parity_synth.py

```diff
@@ -0,0 +1,532 @@
+        if not, and insert its output qubits in ``self.phase_polynomial_qubits``."""
+        for i, q in enumerate(op.in_qubits):
+            if q in self.phase_polynomial_qubits:
+                self.phase_polynomial_qubits.remove(q)
+            else:
+                self.init_wire_map[q] = self.num_phase_polynomial_qubits
+                self.num_phase_polynomial_qubits += 1
+            self.phase_polynomial_qubits.add(op.out_qubits[i])
+        self.phase_polynomial_ops.append(op)
+
+    @pattern_rewriter.op_type_rewrite_pattern
+    def match_and_rewrite(self, matchedOp: Operation, rewriter: pattern_rewriter.PatternRewriter):
+        r"""Implementation of rewriting ``FuncOps`` that may contain phase poynomials
+        with ``ParitySynth``.
+
+        Args:
+            funcOp (func.FuncOp): function containing the operations to rewrite.
+            rewriter (pattern_rewriter.PatternRewriter): Rewriter that executes operation erasure
+                and insertion.
+
+        The logic of this implementation is centered around :attr:`~.rewrite_phase_polynomial`,
+        which is able to rewrite a collection of ``CustomOp``\ s that forms a phase polynomial
+        (see ``valid_phase_polynomial_ops`` for the supported types) into a new collection of
+        ``CustomOp``\ s that is equivalent. In addition to the operators, which are stored in
+        ``self.phase_polynomial_ops``, the ``rewrite_phase_polynomial`` subroutine requires
+        the initial mapping from input qubits to integer-valued wire positions, which is computed
+        in ``self.init_wire_map`` using temporary variables ``self.phase_polynomial_qubits``
+        and ``self.num_phase_polynomial_qubits``.
+
+        Iterating over all operations, the collected phase polynomial ops are rewritten as soon
+        as a non-phase-polynomial operation is encountered. Note that this makes the (size of the)
+        rewritten phase polynomials dependent on the order in which we walk over the operations.
+        """
+        # The attribute is used so we don't transform the same op multiple times
+        if len(matchedOp.regions) == 0 or hasattr(matchedOp, "parity_synth_done"):
+            return
+
+        for region in matchedOp.regions:
+            for block in region.blocks:
+                for op in block.ops:
+                    if not isinstance(op, CustomOp):
+                        if len(op.regions) != 0:
+                            # Do phase polynomial rewriting up to this point
+                            self.rewrite_phase_polynomial(rewriter)
+                            # Rewrite regions of this operation; Creating a new PatternRewriter
+                            # so its matched operation is `op`, not `matchedOp`
+                            self.match_and_rewrite(op, pattern_rewriter.PatternRewriter(op))
+                            op.attributes["parity_synth_done"] = builtin.UnitAttr()
```

Comment:
Just realized this shouldn't be needed. The addition of the attribute later in the function should cover it.
```suggestion
```